src/emmy/expression/compile.cljc

```diff
@@ -223,25 +234,147 @@
+               :flat [(into [] (flatten state-model))]
+               :native state-model
+               (throw (ex-info "Invalid calling convention supplied"
+                               {:calling-convention calling-convention})))]
+    (assoc code
+           :argv (if (and params (not= calling-convention :primitive))
+                   (conj argv (into [] params))
+                   argv))))
+
+(defn- append-bindings
+  "Returns a new let expression with the supplied `bindings`
+  wrapping `exp`. If `exp` is already a let expression, the
+  `bindings` will be appended (or prepended, if `prepend?`)
+  to the existing ones. Each binding
+  is a two-element list [var value]."
+  [exp bindings & prepend?]
+  (let [z (z/seq-zip exp)
+        unpaired-bindings (for [kv bindings a kv] a)]
+    (z/root (if (= (z/node (z/next z)) `let)
+              (z/edit (z/right (z/next z)) (fn [existing-bindings]
+                                             (if prepend?
+                                               (into [] (concat unpaired-bindings existing-bindings))
+                                               (into existing-bindings unpaired-bindings))))
+              (z/replace z `(let ~(vec unpaired-bindings) ~(z/node z)))))))
+
+(defn- cse
+  "Invokes [[emmy.expression.cse/extract-common-subexpressions]] on `x`.
+
+  Local bindings for the common sub-expressions found are appended to the
+  binding structure of the body, and the (possibly) simplified
+  code replaces that in the previous code body."
+  [x gensym-fn deterministic?]
+  (extract-common-subexpressions
+   (:body x)
+   (fn [new-body new-locals]
+     (assoc x :body (if (seq new-locals)
+                      (append-bindings new-body new-locals)
+                      new-body)))
+   {:gensym-fn gensym-fn :deterministic? deterministic?}))
+
+(defn- primitive-bindings
+  "In the primitive calling convention case, introduces a sequence of
+  local variables for each of the state variables, retrieving the values
+  from the primitive input array. Returns an updated code object."
+
+  [{:keys [argv calling-convention state-model params] :as code}]
+  (letfn [(local-vars-from-array
+           ;; Arranges for the local variables given in `vars` to be bound,
+           ;; one by one, to the elements of the primitive array denoted by
+           ;; `array-symbol `.
+           [array-symbol vars]
+           (map-indexed (fn [i v] [v `(aget ~array-symbol ~i)]) vars))
+          ]
+    (case calling-convention
+      :primitive (let [[ys _ ps] argv]
+                   (update code :body append-bindings
+                           (concat (local-vars-from-array ys (flatten state-model))
+                                   (local-vars-from-array ps params))
+                           :prepend))
+      code)))
+
+(defn- primitive-body
+  "In the case of primitive calling convention, unrolls the source of the
+  vector-producing statement of the body into a sequence of `aset-double`
```

Comment:
see `aset-double` note above...

Reply:
you're completely right about `aset-double` and I'm removing it everywhere